geomag_imp/secs.py

```diff
@@ -518,146 +522,141 @@ def make_T(self, lat_lon_r):
-                np.cos(x1) * TZ) # r-component
-         TX1 = (np.cos(x1) * np.cos(x2) * TX +
-                np.cos(x1) * np.sin(x2) * TY -
-                np.sin(x1) * TZ) # poloidal component
-         TX2 = -np.sin(x2) * TX + np.cos(x2) * TY # toroidal component
-
-         # construct output array
-         # (this differs from Pulkkinen's code, and is more like the literature)
-         T_out[0::3, n_el] = TX1
-         T_out[1::3, n_el] = TX2
-         T_out[2::3, n_el] = TX3
-
-      return T_out
+      nobs = len(lat_lon_r)
+      nsec = len(self.secs.lat_lon_r)
+
+      obs_r = lat_lon_r[:, 2][:, np.newaxis]
+      sec_r = self.secs.lat_lon_r[:, 2][np.newaxis, :]
+
+      theta = calc_angular_distance(lat_lon_r[:, :2], self.secs.lat_lon_r[:, :2])
+      alpha = calc_bearing(lat_lon_r[:, :2], self.secs.lat_lon_r[:, :2])
+
+      # magnetic permeability
+      mu0 = 4*np.pi*1e-7
+
+      # simplify calculations by storing this ratio
+      x = obs_r/sec_r
+
+      sin_theta = np.sin(theta)
+      cos_theta = np.cos(theta)
+      factor = 1./np.sqrt(1 - 2*x*cos_theta + x**2)
+
+      # Amm & Viljanen: Equation 9
+      Br = mu0/(4*np.pi*obs_r) * (factor - 1)
+
+      # Amm & Viljanen: Equation 10 (transformed to try and eliminate trig 
+      #                              operations and divide by zeros)
+      Btheta = -mu0/(4*np.pi*obs_r) * (factor*(x - cos_theta) + cos_theta)
+      # If sin(theta) == 0: Btheta = 0
+      # There is a possible 0/0 in the expansion when sec_loc == obs_loc
+      Btheta = np.divide(Btheta, sin_theta, out=np.zeros_like(sin_theta),
+                        where=sin_theta != 0)
+
+      # When observation points radii are outside of the sec locations
+      under_locs = sec_r < obs_r
+
+      # NOTE: If any SECs are below observations the math will be done on all
+      #       points. This could be updated to only work on the locations where
+      #       this condition occurs, but would make the code messier, with 
+      #       minimal performance gain except for very large matrices.
+      if np.any(under_locs):
+         # Flipped from previous case
+         x = sec_r/obs_r
+
+         # Amm & Viljanen: Equation A.7
+         Br2 = mu0*x/(4*np.pi*obs_r) * (1./np.sqrt(1 - 2*x*cos_theta + x**2) - 1)
+
+         # Amm & Viljanen: Equation A.8
+         Btheta2 = - mu0 / (4*np.pi*obs_r) * ((obs_r-sec_r*cos_theta) /
+                                                np.sqrt(obs_r**2 -
+                                                      2*obs_r*sec_r*cos_theta +
+                                                      sec_r**2) - 1)
+         Btheta2 = np.divide(Btheta2, sin_theta, out=np.zeros_like(sin_theta),
+                              where=sin_theta != 0)
+
+         # Update only the locations where secs are under observations
+         Btheta[under_locs] = Btheta2[under_locs]
+         Br[under_locs] = Br2[under_locs]
+
+      # Transform back to Bx, By, Bz at each local point
```

Comment:
After our discussion, you may want to update this to say Btheta, Bphi, Br instead?